src/server/serve.rs

```diff
@@ -393,37 +454,54 @@ impl InnerService {
             }
         }
 
-        let mut body = body?;
         let mime_type = InnerService::guess_path_mime(&path, action);
 
-        if self.can_compress(res.status(), &mime_type) {
-            let encoding = res
+        let body = if self.can_compress(res.status(), &mime_type) {
+            let encoding = req
                 .headers()
                 .get(hyper::header::ACCEPT_ENCODING)
                 .map(get_prior_encoding)
                 .unwrap_or_default();
-            // No Accept-Encoding would not be compress.
-            if let Ok(buf) = compress(&body, encoding) {
-                body = buf;
-                res.headers_mut().insert(
-                    hyper::header::CONTENT_ENCODING,
-                    hyper::header::HeaderValue::from_static(encoding),
-                );
-                // Representation varies, so responds with a `Vary` header.
-                res.headers_mut().insert(
-                    hyper::header::VARY,
-                    hyper::header::HeaderValue::from_name(hyper::header::ACCEPT_ENCODING),
-                );
+            let enc = encoding_to_static_str(encoding);
+            match enc {
+                BR | DEFLATE | GZIP => {
```

Comment:
I'd like to wrap compression stream as `hyper::Body` but don't want to leak `hyper::Body` to `src/http/content_encoding.rs`. Maybe there is better way to achieve?

Reply:
I am thinking of a `fn should_compress(enc: &str) -> bool` doing that job, so that either `hyper::Body` or encoding string won't leak to elsewhere.

Inside `content_encoding::should_compress` it just checks if we got an `IDENTITY` encoding. WDYT?

Reply:
agree https://github.com/weihanglo/sfz/pull/97#discussion_r963865662, I will implement `content_encoding::should_compress`

Reply:
Since BR, DEFLATE, GZIP are encapsulated in `content_encoding.rs` after the implementation, I will move them from `http/mod.rs` back to `http/content_encoding.rs`.

Reply:
I think we should explicitly check whether `Accept-Encoding` is `br`, `deflate` or `gzip` because we currently support these algorithms. If client send other algorithm such as LZMA, we should ignore.

Reply:
> I think we should explicitly check whether Accept-Encoding is br, deflate or gzip because we currently support these algorithms.

Already did that inside `get_prior_encoding`[^1]. Somehow obscure though.

[^1]: https://github.com/weihanglo/sfz/blob/7113041be5f1719b4aef89f3bd1b2dc0290d04ca/src/http/content_encoding.rs#L101-L123

Reply:
I add a `dbg!(encoding, should_compress(encoding));` on https://github.com/weihanglo/sfz/blob/7113041be5f1719b4aef89f3bd1b2dc0290d04ca/src/server/serve.rs#L405

curl

```
$ curl -v http://localhost:5000/Cargo.toml
*   Trying 127.0.0.1:5000...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 5000 (#0)
> GET /Cargo.toml HTTP/1.1
> Host: localhost:5000
> User-Agent: curl/7.68.0
> Accept: */*
>
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
```

sfz

```
Files served on http://127.0.0.1:5000
[src/server/serve.rs:416] encoding = ""
[src/server/serve.rs:416] should_compress(encoding) = true
[07/Sep/2022 23:19:14] "GET /Cargo.toml" - 200 OK
```

should_compress

```rust
pub fn should_compress(enc: &str) -> bool {
    IDENTITY != enc // it will be true because empty string never equals to "identity"
}
```

Reply:
I created a simple playground to describe the issue https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1c8c1b1f3c521307c34374ffb4989e89

Reply:
You observation is totally correct. Need to modify the surrounding context accordingly. Something like making `get_prior_encoding` accept `Option<&HeaderValue>` and avoid unwrap_or_default.

Reply:
Implement `should_compress` and fix `get_prior_encoding` usage on 5a1ed98